### ISSUE TITLE
Make Compose invocations reentrant by removing shared call state

### ISSUE THREAD

**Issue by ternaus** [albumentations-team/AlbumentationsX#419]
## Summary

`Compose` stores invocation-specific bookkeeping on the configured pipeline and transform instances. Two overlapping calls to the same `Compose` object can overwrite that bookkeeping and let one sample finish with state belonging to another sample.

This issue defines the concurrency contract for `Compose` and removes the shared invocation state that violates it. PR #418 exposed the problem while adding a Tensor-to-NumPy compatibility route for mixed transform pipelines. The underlying limitation already predates that PR.

## Why this matters

The common `DataLoader(num_workers > 0)` configuration usually creates a separate dataset and `Compose` instance in each worker process. It does not share one Python object and therefore does not exercise this race.

The race appears when application code shares one configured pipeline across overlapping calls, for example through a `ThreadPoolExecutor`, a web service, a threaded data source, or `asyncio.to_thread`. CPython's GIL does not make a multi-step mutation/read/clear protocol atomic. NumPy and Torch operations can also run native work while the GIL is released.

The desired property is reentrancy: overlapping calls to the same `Compose` instance must produce results consistent with some serial order. Each result must keep its own representation, layout, annotations, applied configuration, and trace. A shared seed does not need to promise which caller receives the next random draw unless the caller imposes an ordering.

## Concrete race exposed by #418

PR #418 adds three fields that describe the current Tensor call:

- `_tensor_spatial_targets`
- `_tensor_annotation_targets`
- `_tensor_requires_numpy_bridge`

For a Tensor input whose pipeline contains a NumPy-only transform, `Compose._validate_tensor_inputs()` records the supplied targets and selects the one whole-pipeline NumPy bridge. `_bridge_tensor_data_to_numpy()` and the restoration helpers read those fields later in the same invocation.

The following interleaving is valid today:

```text
call A: Tensor image + Tensor mask enter a pipeline that needs the NumPy bridge
call A: records targets=(image, mask), requires_numpy_bridge=True

call B: NumPy image, or a Tensor call with only image, enters the same Compose object
call B: clears or replaces the three fields

call A: resumes and reads B's state
```

Call A can then skip its required conversion, convert only its image while leaving its mask as a Tensor, or skip conversion back to Tensor at the public output boundary. A NumPy-only transform can receive a Tensor, and an otherwise valid call can return inconsistent target types.

`try/finally` cleanup prevents state from leaking after a serial call raises. It cannot isolate two calls whose lifetimes overlap.

## Existing shared invocation state

The Tensor route is one instance of a broader pattern. The current code also stores these per-invocation values on `Compose`:

| State | Used for | Failure when another call replaces it |
| --- | --- | --- |
| `_added_channel_dim`, `_added_channel_canonical` | Restoring a grayscale target after preprocessing | A call can retain an added channel or remove one that belonged to another call. |
| `_instance_count`, `_repack_after_processors` | Rebuilding `instance_binding` outputs | A call can repack against another call's instance count or cleanup flags. |
| `_tensor_spatial_targets`, `_tensor_annotation_targets`, `_tensor_requires_numpy_bridge` | Choosing and undoing the Tensor compatibility bridge | A call can use another call's conversion plan and output type contract. |

`BasicTransform` has a related runtime-state problem. Each invocation writes `self.params` and `self.applied_config`; `Compose` reads `applied_config` after the transform returns when `save_applied_params=True`. Concurrent calls can therefore associate one sample's applied configuration with another sample. `run_with_trace()` has the same concern when it records transform state.

All configured transforms also share the pipeline's RNG objects. Sharing an RNG stream can be acceptable if calls are serialized and the contract permits the caller-dependent order of draws. It needs an explicit policy before claiming parallel execution or per-call reproducibility.

## Required scope and decision

The implementation must choose and document one of these contracts:

1. **Serialized concurrent calls.** Guard each outer `Compose.__call__()` and `Compose.run_with_trace()` invocation with a reentrant lock. This preserves the existing mutable pipeline model and produces a valid serial order. It does not provide parallel augmentation from one pipeline object.
2. **Fully isolated invocation contexts.** Pass a call-local context through `Compose`, nested compositions, processors, tracing, and transforms. The context owns representation planning, grayscale restoration, instance-binding bookkeeping, applied parameters, and RNG allocation. This enables genuine overlap, but it is a larger API and state-management redesign.

The first implementation may serialize calls. It must not present that as parallel augmentation. Callers that need parallel CPU work can use independent pipeline instances until an isolated-context design exists.

Regardless of the selected contract, PR #418 should keep its Tensor bridge plan in a call-local value, for example `TensorCallState`, rather than add another observable lifetime-dependent field to `Compose`. That targeted change prevents the PR from adding a new race while this issue tracks the whole object model.

## Implementation plan

1. Inventory every field written during `__call__()` or `run_with_trace()` on `Compose`, nested compositions, processors, and `BasicTransform`.
2. Classify each field as immutable configuration, reusable cache, random state, or invocation state.
3. Move the Tensor bridge plan in #418 to a local object passed explicitly to bridge and restore helpers.
4. Implement the selected concurrency contract across both normal and traced execution. The synchronization boundary must cover RNG synchronization, parameter sampling, transform application, bookkeeping, postprocessing, restoration, and cleanup.
5. Preserve serial behavior, serialization, replay, `save_applied_params`, additional targets, grayscale handling, and instance binding.
6. Document the public concurrency and seeded-RNG semantics in the `Compose` API and the Tensor CPU backend design document.

## Regression coverage

Add permanent tests that use a deterministic synchronization mechanism. A timing-based `sleep` test is insufficient.

- Run overlapping calls to the same `Compose` instance with Tensor fallback inputs whose target sets differ, such as `image + mask` and `image`.
- Assert that every caller receives its own expected Tensor/NumPy boundary type and spatial layout.
- Cover `__call__()` and `run_with_trace()`.
- Cover grayscale channel restoration and `instance_binding` bookkeeping, because they use existing shared invocation state.
- With `save_applied_params=True`, verify that each output records parameters sampled for that exact input.
- Define the seeded-RNG assertion from the selected contract. A serialized implementation should test a valid serial sequence; an isolated implementation should test the documented per-call stream rule.
- Run a focused threaded regression test plus the relevant Compose, Tensor, replay, tracing, and DataLoader suites. The DataLoader test verifies integration; it does not replace the shared-object threaded test.

## Completion criteria

- The public docs state whether one `Compose` object supports overlapping calls and define seeded behavior under that use.
- No invocation state can be overwritten by another overlapping invocation.
- Tensor fallback, grayscale restoration, instance binding, `save_applied_params`, and tracing satisfy the chosen contract.
- Regression tests fail against the vulnerable behavior and pass deterministically under the selected implementation.
- Existing serial Compose behavior and public input/output contracts remain unchanged.

## Out of scope

This issue does not add CUDA, MPS, autograd, or parallel kernel execution to the CPU Tensor backend. It also does not treat separate `DataLoader` worker processes as concurrent access to one Python `Compose` object.

**Comment by ternaus** [comment]
Fixed in https://github.com/albumentations-team/AlbumentationsX/pull/422
